app/notification_push_test.go

```diff
@@ -1341,3 +1348,132 @@ func TestAllPushNotifications(t *testing.T) {
 	assert.Equal(t, 6, numUpdateBadges)
 }
+
+func BenchmarkPushNotification(b *testing.B) {
+	th := SetupWithStoreMock(b)
+	defer th.TearDown()
+
+	handler := &testPushNotificationHandler{
+		t:        b,
+		behavior: "simple",
+	}
+	pushServer := httptest.NewServer(
+		http.HandlerFunc(handler.handleReq),
+	)
+	defer pushServer.Close()
+
+	mockStore := th.App.Srv().Store.(*mocks.Store)
+	mockUserStore := mocks.UserStore{}
+	mockUserStore.On("Count", mock.Anything).Return(int64(10), nil)
+	mockUserStore.On("GetUnreadCount", mock.AnythingOfType("string")).Return(int64(1), nil)
+	mockPostStore := mocks.PostStore{}
+	mockPostStore.On("GetMaxPostSize").Return(65535, nil)
+	mockSystemStore := mocks.SystemStore{}
+	mockSystemStore.On("GetByName", "InstallationDate").Return(&model.System{Name: "InstallationDate", Value: "10"}, nil)
+	mockSessionStore := mocks.SessionStore{}
+	mockPreferenceStore := mocks.PreferenceStore{}
+	mockPreferenceStore.On("Get", mock.AnythingOfType("string"), mock.AnythingOfType("string"), mock.AnythingOfType("string")).Return(&model.Preference{Value: "test"}, nil)
+	mockStore.On("User").Return(&mockUserStore)
+	mockStore.On("Post").Return(&mockPostStore)
+	mockStore.On("System").Return(&mockSystemStore)
+	mockStore.On("Session").Return(&mockSessionStore)
+	mockStore.On("Preference").Return(&mockPreferenceStore)
+
+	// create 50 users, each having 2 sessions.
+	type userSession struct {
+		user    *model.User
+		session *model.Session
+	}
+	var testData []userSession
+	for i := 0; i < 50; i++ {
+		id := model.NewId()
+		u := &model.User{
+			Id:            id,
+			Email:         "success+" + id + "@simulator.amazonses.com",
+			Username:      "un_" + id,
+			Nickname:      "nn_" + id,
+			Password:      "Password1",
+			EmailVerified: true,
+		}
+		sess1 := &model.Session{
+			Id:        "id1",
+			UserId:    u.Id,
+			DeviceId:  "deviceID" + u.Id,
+			ExpiresAt: model.GetMillis() + 100000,
+		}
+		sess2 := &model.Session{
+			Id:        "id2",
+			UserId:    u.Id,
+			DeviceId:  "deviceID" + u.Id,
+			ExpiresAt: model.GetMillis() + 100000,
+		}
+		mockSessionStore.On("GetSessionsWithActiveDeviceIds", u.Id).Return([]*model.Session{sess1, sess2}, nil)
+		mockSessionStore.On("UpdateDeviceId", sess1.Id, "deviceID"+u.Id, mock.AnythingOfType("int64")).Return("deviceID"+u.Id, nil)
+
+		testData = append(testData, userSession{
+			user:    u,
+			session: sess1,
+		})
+	}
+
+	th.App.UpdateConfig(func(cfg *model.Config) {
+		*cfg.EmailSettings.PushNotificationServer = pushServer.URL
+		*cfg.LogSettings.EnableConsole = false
+		*cfg.NotificationLogSettings.EnableConsole = false
+	})
+
+	ch := &model.Channel{
+		Id:       model.NewId(),
+		CreateAt: model.GetMillis(),
+		Type:     model.CHANNEL_OPEN,
+		Name:     "testch",
+	}
+
+	b.ResetTimer()
+	// We have an inner loop which ranges the testdata slice
+	// and we just repeat that.
+	// TODO: replace 10 by b.N
+	then := time.Now()
+	for i := 0; i < 10; i++ {
```

Comment:
I read the comment but I'm wondering why didn't you start using the b.N in the for loop ?

Reply:
Oh .. sorry. Yes, this is because the code is not even efficient enough to run till b.N. Since we are creating a new http.Client for every single call, it runs out of file descriptors before even finishing b.N iterations.

That's my next optimization which I have in the pipeline. There this TODO is resolved. I will send it once these 3 are merged.